Remove local functionality

msimberg opened the pull request:
Companion PR to #5578. Removes local-only functionality, and replaces it with a dependency on `HPXLocal`. Work in progress...

msimberg commented:
I'd appreciate reviews on this now. I'll try to summarize the changes, but please ask if you think I haven't explained something in enough detail (or if it seems like I've forgotten something). If you see problems with the choices here, please let me know. Nothing is set in stone.

- This removes the `libs/core` directory and the `hpx_core` target from this repository.
- It adds an option `HPX_WITH_FETCH_HPXLOCAL` to automatically fetch `HPXLocal`. This is on by default to be as similar to the current setup with a single repo as possible.
- Most CMake options have been separated to be either for this repo (still called `HPX_WITH_...`) or for HPXLocal (called `HPXLocal_WITH_...`).
- When fetching HPXLocal `HPX_WITH_...` variables that were moved to HPXLocal will be forwarded to `HPXLocal_WITH_...` variables with a warning.
- When HPXLocal is installed separately no forwarding or warnings will happen as HPXLocal is already built.
- All C++ feature tests happen only in HPXLocal, and HPX reuses their results (this may need to be revisited if we think the two will have a need for significantly different feature tests).
- The C++ standard is now set in HPXLocal with `HPXLocal_CXX_STANDARD`. This replaces the use of `CMAKE_CXX_STANDARD` and `HPX_WITH_CXXAB`. When I made the change to prefer `CMAKE_CXX_STANDARD` I did not take into account the possibility that we'd need `CMAKE_CUDA_STANDARD` (and possibly `CMAKE_HIP_STANDARD` in the future), so it makes more sense to have an HPX-specific variable for it.
- The `HPX_WITH_CXXAB` variables forward to `HPXLocal_CXX_STANDARD` with a warning if HPXLocal is fetched.
- Documentation lives entirely in this repository. The documentation will include API documentation for HPXLocal if HPXLocal is fetched.
- I've removed the CUDA options `HPX_WITH_COMPUTE_CUDA` and `HPX_WITH_ASYNC_CUDA` since one of the modules is now in HPX and the other in HPXLocal. There's only `HPXLocal_WITH_CUDA` (and `HPXLocal_WITH_HIP`) now. If `HPXLocal_WITH_CUDA=ON`, HPX will also use CUDA.
- Likewise, there's no `HPX_WITH_ASYNC_MPI` option anymore. There is `HPXLocal_WITH_MPI`, and if `HPXLocal_WITH_MPI=ON` HPX will automatically set `HPX_WITH_PARCELPORT_MPI=ON`.
- There is no `HPX_WITH_DISTRIBUTED_RUNTIME` option anymore. HPXLocal only provides the local runtime entry point. Applications compiled with HPX can still choose to use the local-only runtime at compile- or run-time (with `--hpx:local`) as before.
- CMake options affecting the build system (`UNITY_BUILD`, `PRECOMPILED_HEADERS`) are separate for HPX and HPXLocal.
- Tests and examples for HPXLocal are off by default, and will not be built or run when fetching HPXLocal. They can be enabled and built through the same targets, however (i.e. `tests`, `tests.unit`, `examples`, etc.).
- Until the next release HPXLocal is fixed to a specific tag/commit (currently the branch corresponding to https://github.com/STEllAR-GROUP/hpx-local/pull/1) and should be updated explicitly. HPXLocal is set to version 0.0.1 at the moment and the exact version is required by the CMake configuration.
- After the next release I propose we still require an exact version, but only released versions of HPXLocal. Once things stabilize, we may require only a minimum version.

aurianer commented on .jenkins/cscs/env-gcc-oldest.sh:
```diff
@@ -23,15 +23,14 @@ module load daint-mc
+configure_extra_options+=" -DHPXLocal_WITH_MALLOC=system"
+configure_extra_options+=" -DHPXLocal_WITH_CXX${CXX_STD}=ON"
+configure_extra_options+=" -DHPXLocal_WITH_GENERIC_CONTEXT_COROUTINES=ON"
+configure_extra_options+=" -DHPXLocal_WITH_APEX=ON"
```
```suggestion
configure_extra_options+=" -DHPX_WITH_APEX=ON"
```
? cause in the env-clang-apex.sh, this one is stated

aurianer commented on .jenkins/lsu/env-clang-9.sh:
```diff
@@ -14,15 +14,14 @@ module load openmpi
+configure_extra_options+=" -DHPXLocal_WITH_CXX${CXX_STD}=ON"
+configure_extra_options+=" -DHPXLocal_WITH_MALLOC=system"
+configure_extra_options+=" -DHPXLocal_WITH_LOGGING=OFF"
+configure_extra_options+=" -DHPXLocal_WITH_ASYNC_MPI=ON"
```
```suggestion
configure_extra_options+=" -DHPXLocal_WITH_MPI=ON"
```
?

msimberg replied:
See: https://github.com/STEllAR-GROUP/hpx-local/pull/2#discussion_r736682825. If there's no preference either way I'll change the variable to `HPXLocal_WITH_MPI`.

hkaiser replied:
Why is the MPIAsync stuff part of the local repository now. That doesn't make any sense to me :/

msimberg replied:
Because of a few reasons. Partly because of the current integration with the scheduler (exactly as `async_cuda`). `async_mpi` is neither distributed HPX nor local HPX and I want to separate this out into a separate repository, which I'm pretty sure I've mentioned before. However, that needs a different mechanism for registering the polling functionality, and I don't want to do that at the same time as this split. Since users of the _real_ distributed HPX would have very little reason to use the `async_mpi` module it makes sense to have it here at this point.

msimberg replied:
Note that this means that the only thing distributed HPX needs from local HPX is really the `mpi_base` module, which can eventually stay here, move to distributed HPX or be duplicated in a future `hpx-async-mpi` library.

hkaiser replied:
Well, as said - having anything related to MPI in a repository that positions itself as 'local' doesn't sound very sensible to me. The arguments you list (integration with scheduler, registration of polling functions, future plans to separate it, etc.) sound orthogonal to the issue. All of those arguments were equally true and applicable if the MPIAsync module stayed in the distributed part  (where I believe it belongs). 

For the claim that users of distributed HPX have little reason to use MPIAsync - I would beg to disagree. One of our projects (FleCSI) has a diverse set of use cases where we might have to use both, HPX collectives and MPIAsync functionalities at the same time, at least for evaluation purposes.

msimberg replied:
I think my comment about it belonging eventually in a separate repository agrees with your comment that HPXLocal is not the right place long-term. There are technical reasons for having it there now, but my preference is also not to have it there long term.

HPXLocal provides a local-only runtime. That does not mean it can't be used in distributed settings. `async_mpi` is a lightweight wrapper to use asynchronous MPI functions in a runtime that otherwise doesn't know anything about distributed computing.

Finally, you know that we rely on functionality in `async_mpi`. Having it in the distributed HPX means that we will keep depending on the full distributed HPX, which is counter to this whole split. Having it in the local or distributed HPX makes zero difference to FleCSI.

I want to repeat: long-term it doesn't belong in this repository, but it doesn't belong in distributed HPX either. It's an alternative way to do distributed computing in HPX without the distributed runtime.

hkaiser replied:
Let's discuss this in person at some point. I'm not having a good feeling about where this is going...

msimberg replied:
Agreed. Let's do that.

msimberg replied:
We discussed this yesterday and I think we're on the same page now. We'll keep `async_mpi` here for the initial split and try to put a more generic mechanism in place in the schedulers to have the MPI polling register itself from a separate library afterwards.

msimberg replied:
Also, regarding the initial question, I'll keep the name since it'll be removed once the functionality is in a separate repository (https://github.com/STEllAR-GROUP/hpx-local/pull/2#discussion_r748137825).

aurianer reviewed (commented):
Thanks a lot for working on this!  Same I will do my review in several parts cause it's quite long ^^'

msimberg commented on .circleci/config.yml:
```diff
@@ -204,27 +204,28 @@ jobs:
+                -DCMAKE_CXX_FLAGS="-fcolor-diagnostics" \
+                -DCMAKE_CXX_CLANG_TIDY=clang-tidy \
+                -DCMAKE_EXPORT_COMPILE_COMMANDS=On \
+                -DHPX_WITH_FETCH_HPXLOCAL=On \
```
See https://github.com/STEllAR-GROUP/hpx-local/pull/2#discussion_r735383894 for a motivation. We can go either way here as well. If there's no preference either way, I'll leave it as it is.

aurianer replied:
Sorry forgot about that ^^

aurianer commented on CMakeLists.txt:
```diff
@@ -498,543 +430,40 @@ hpx_option(
+  ON
   ADVANCED
 )
-if(HPX_WITH_NICE_THREADLEVEL)
```
Should we keep it for `HPXLocal_WITH_NICE_THREADLEVEL`?

msimberg replied:
No, the definition is added in HPXLocal.

aurianer reviewed (commented):
A few more comments

aurianer commented on CMakeLists.txt:
```diff
@@ -1939,59 +1094,6 @@ else()
-  include(TargetArch)
-
-  target_architecture(__target_arch)
-  if("${__target_arch}" STREQUAL "i386"
```
4 occurrences of `__target_arch` are still here in the `CMakeLists.txt`, I'm not sure it was intended

aurianer commented on CMakeLists.txt:
```diff
@@ -498,543 +430,40 @@ hpx_option(
-   OR HPX_WITH_HIP
-)
-  set(HPX_WITH_PKGCONFIG_DEFAULT OFF)
-endif()
```
Actually just noticed the cache force later, ignore my comment :)

msimberg replied:
As you noted somewhere (can't find where...) I now force this to off on Windows and with GPU support enabled: https://github.com/msimberg/hpx/blob/2f163ce1bf2a3090f6fd581aaa5ac5537e02ef70/CMakeLists.txt#L1146-L1154.

msimberg commented on .jenkins/cscs/env-clang-12.sh:
```diff
@@ -23,19 +23,19 @@ module load daint-mc
 configure_extra_options+=" -DHPX_WITH_COMPILER_WARNINGS=ON"
 configure_extra_options+=" -DHPX_WITH_COMPILER_WARNINGS_AS_ERRORS=ON"
-configure_extra_options+=" -DHPX_WITH_SPINLOCK_DEADLOCK_DETECTION=ON"
 configure_extra_options+=" -DHPX_WITH_UNITY_BUILD=ON"
```
`env-common.sh` had `HPXLocal_WITH_UNITY_BUILD=ON` but I've added both to `env-common.sh` now.

msimberg commented on CMakeLists.txt:
```diff
@@ -498,543 +430,40 @@ hpx_option(
-# HPX datapar configuration
-# ##############################################################################
-hpx_option(
-  HPX_WITH_DATAPAR_VC
```
`HPXLocal_WITH_DATAPAR_VC` is in `.circleci/config.yml` and I'm planning to leave it there. Good catch on the others though! Will fix.

msimberg replied:
And I just noticed the other one in `.circleci/config.yml`... thanks!

msimberg commented on CMakeLists.txt:
```diff
@@ -498,543 +430,40 @@ hpx_option(
+  HPX_WITH_FETCH_HPXLOCAL
   BOOL
-  "Enable keeping track of cumulative thread counts in the schedulers (default: ON)"
+  "Use FetchContent to fetch HPXLocal. By default an installed HPXLocal will be used. (default: OFF)"
```
Good catch!

msimberg replied:
By the way, the correct one is the actual default, not the one in the string.

aurianer commented on tests/performance/local/CMakeLists.txt:
```diff
@@ -7,25 +7,6 @@
-)
-
 set(timed_task_spawn_SOURCES activate_counters.cpp)
 set(timed_task_spawn_HEADERS activate_counters.hpp)
```
Looks like this benchmark has been removed

aurianer commented on tests/performance/local/CMakeLists.txt:
```diff
@@ -34,93 +15,24 @@ if(NOT HPX_WITH_CUDA_COMPUTE)
   set(stream_FLAGS CUDA)
```
Couldn't point the specific line but there is still a `HPX_WITH_CUDA_COMPUTE` (two lines above)

aurianer commented on tests/performance/local/CMakeLists.txt:
```diff
@@ -34,93 +15,24 @@ if(NOT HPX_WITH_CUDA_COMPUTE)
-set(hpx_homogeneous_timed_task_spawn_executors_FLAGS DEPENDENCIES
-                                                     iostreams_component
-)
 set(skynet_FLAGS DEPENDENCIES iostreams_component)
```
same this one is gone I believe

aurianer commented on libs/full/version/src/version.cpp:
```diff
@@ -283,9 +265,9 @@ namespace hpx {
-            build_string(), boost_version(), hwloc_version(),
-#if (defined(HPX_HAVE_NETWORKING) && defined(HPX_HAVE_PARCELPORT_MPI)) ||      \
-    defined(HPX_HAVE_MODULE_MPI_BASE)
+            build_string(), hpx::build_string(), boost_version(),
```
```suggestion
            build_string(), hpx::local::build_string(), boost_version(),
```
?

msimberg replied:
Yep. Good catch!

aurianer commented on libs/full/version/CMakeLists.txt:
```diff
@@ -10,20 +10,13 @@ set(version_headers hpx/version.hpp)
-  MODULE_DEPENDENCIES hpx_config hpx_config_registry hpx_format hpx_prefix
-  DEPENDENCIES Hwloc::hwloc ${mpi_dependencies}
-  MODULE_DEPENDENCIES hpx_config hpx_format hpx_prefix
+  MODULE_DEPENDENCIES hpx_config
```
Isn't format still a dependency here?

msimberg replied:
No. `hpx_format` is a module in `HPX::hpx_local`. The fact that it's a module in another library is something that `HPX::hpx` does not know about. Linking to it here would lead to duplicated symbols.

aurianer reviewed (commented):
A few more comments

msimberg commented on cmake/templates/global_module_header.hpp.in:
```diff
@@ -10,5 +10,6 @@
 
-#if defined(HPX_HAVE_MODULE_@modulename_upper@)
-@module_headers@#endif
+//#if defined(HPX_HAVE_MODULE_@modulename_upper@)
```
Good catch. Yes, I intended to remove this guard as I don't see the use of it. If the module isn't enabled, the whole file won't exist.

msimberg commented on tools/inspect/CMakeLists.txt:
```diff
@@ -19,7 +19,7 @@ endif()
 target_include_directories(inspect PRIVATE ${PROJECT_BINARY_DIR})
 target_link_libraries(
-  inspect PRIVATE Boost::regex hpx_dependencies_boost hpx_core
+  inspect PRIVATE Boost::regex hpx_dependencies_boost HPX::hpx
```
Yes, that would be enough.

msimberg replied:
I take this back. It adds the commit hash which comes from `HPX::hpx`. I've changed it back to `HPX::hpx`.

msimberg commented on tests/unit/util/CMakeLists.txt:
```diff
@@ -4,11 +4,7 @@
-if(HPX_WITH_DISTRIBUTED_RUNTIME)
-  set(tests ${tests} bind_action)
-endif()
+set(tests config_entry bind_action)
```
Yes, I wanted to duplicate it to make sure that it works correctly with both the local and the distributed runtime.

aurianer commented on libs/full/actions/tests/unit/CMakeLists.txt:
```diff
@@ -13,7 +13,6 @@ endif()
 set(set_thread_state_PARAMETERS THREADS_PER_LOCALITY 4)
 set(thread_stacksize_PARAMETERS LOCALITIES 2)
 set(thread_affinity_PARAMETERS THREADS_PER_LOCALITY 4 RUN_SERIAL)
-set(thread_affinity_LIBRARIES DEPENDENCIES PRIVATE Hwloc::hwloc)
```
Was this intentional?

msimberg replied:
Intentional, but not a must. Since HPX already depends on hwloc I figured we don't really need to put this here explicitly. But having it there wouldn't be wrong either.

aurianer commented on libs/full/async_distributed/CMakeLists.txt:
```diff
@@ -119,7 +115,7 @@ add_hpx_module(
   COMPAT_HEADERS ${async_compat_headers}
-  DEPENDENCIES hpx_core
+  DEPENDENCIES HPX::hpx_local
   MODULE_DEPENDENCIES hpx_actions hpx_naming
```
Should it also depend on `hpx_config`?

aurianer commented on libs/full/collectives/CMakeLists.txt:
```diff
@@ -70,7 +66,7 @@ add_hpx_module(
   COMPAT_HEADERS ${collectives_compat_headers}
-  DEPENDENCIES hpx_core
+  DEPENDENCIES HPX::hpx_local
   MODULE_DEPENDENCIES hpx_actions hpx_async_colocated hpx_async_distributed
```
```suggestion
  MODULE_DEPENDENCIES hpx_actions hpx_async_colocated hpx_async_distributed hpx_config
```
?

aurianer reviewed (approved):
Thanks a lot for the 2 PRs!

msimberg commented:
@hkaiser I've added an advanced CMake variable `HPXLocal_WITH_BINARY_DIR` (to HPXLocal) which is used as the root binary output directory. This was previously using `PROJECT_BINARY_DIR` directly as the root. I set it to default to `HPX_BINARY_DIR` when on Windows and HPX has set `HPX_WITH_FETCH_HPXLOCAL=ON`, and otherwise it defaults to the usual `PROJECT_BINARY_DIR`. I'm hoping this will make sure libraries are found correctly on Windows, but I'd appreciate if you could try it out as well. CI will likely tell me before that as well...

I've also removed the duplicate tests and examples that you removed in https://github.com/msimberg/hpx/pull/8. You should be able to drop that commit from https://github.com/msimberg/hpx/pull/8.

hkaiser commented:
> @hkaiser I've added an advanced CMake variable `HPXLocal_WITH_BINARY_DIR` (to HPXLocal) which is used as the root binary output directory. This was previously using `PROJECT_BINARY_DIR` directly as the root. I set it to default to `HPX_BINARY_DIR` when on Windows and HPX has set `HPX_WITH_FETCH_HPXLOCAL=ON`, and otherwise it defaults to the usual `PROJECT_BINARY_DIR`. I'm hoping this will make sure libraries are found correctly on Windows, but I'd appreciate if you could try it out as well. CI will likely tell me before that as well...

I missed this comment before. Thanks for applying this change. I think we should set the `HPXLocal_WITH_BINARY_DIR` variable in the distributed repository just before calling fetch_content, 

> I've also removed the duplicate tests and examples that you removed in [msimberg#8](https://github.com/msimberg/hpx/pull/8). You should be able to drop that commit from [msimberg#8](https://github.com/msimberg/hpx/pull/8).

Will do.

msimberg commented:
> > @hkaiser I've added an advanced CMake variable `HPXLocal_WITH_BINARY_DIR` (to HPXLocal) which is used as the root binary output directory. This was previously using `PROJECT_BINARY_DIR` directly as the root. I set it to default to `HPX_BINARY_DIR` when on Windows and HPX has set `HPX_WITH_FETCH_HPXLOCAL=ON`, and otherwise it defaults to the usual `PROJECT_BINARY_DIR`. I'm hoping this will make sure libraries are found correctly on Windows, but I'd appreciate if you could try it out as well. CI will likely tell me before that as well...
> 
> I missed this comment before. Thanks for applying this change. I think we should set the `HPXLocal_WITH_BINARY_DIR` variable in the distributed repository just before calling fetch_content,

It's now here: https://github.com/msimberg/hpx/blob/1698bc8d61c7210bdfeb897f64a20f80c590f5c9/cmake/HPX_SetupHPXLocal.cmake#L171-L177.

> > I've also removed the duplicate tests and examples that you removed in [msimberg#8](https://github.com/msimberg/hpx/pull/8). You should be able to drop that commit from [msimberg#8](https://github.com/msimberg/hpx/pull/8).
> 
> Will do.

Note, that was already done and merged ;)

msimberg commented:
bors merge
